### ISSUE TITLE
Question about the methods

### ISSUE THREAD

**Issue by DP1701** [lightly-ai/lightly-train#57]
Hello everyone,

First of all, thank you for developing and providing this very helpful and user-friendly library.

While working with the framework, a few questions have come up:
	1.	The [project page](https://docs.lightly.ai/train/stable/methods/distillation.html) mentions that the ViT-B/14 model is used for the distillation method. Where exactly is this model stored on a Linux system?
	2.	Is it possible to use larger models such as ViT-L/14 or ViT-G/14, which have also been [released by Meta](https://github.com/facebookresearch/dinov2), within the distillation method?
	3.	Is there a plan to support the DINOv2 method in the future, or is the distillation approach intended to serve as a replacement?

**Comment by liopeer** [comment]
Hi @DP1701, we are glad that you like working with LightlyTrain!

About your questions:
1. In the versions before 0.6.3 (our current release), we relied on Huggingface's Transformer library to download the checkpoint of the teacher model, this means your the model was in `~/.cache/huggingface`. In 0.6.3 we got rid of Transformers as a dependency and we now manually download the model to `~/.cache/lightly-train`. You can change the default cache directory through the environment variable `LIGHTLY_TRAIN_CACHE_DIR` if desired.
2. You can in fact already use other teacher models however we did not make this public yet. If you would already like to try it out you can specify the teacher model through the `method_args` parameter of the `train` function (however be aware that this might still be subject to change, since we did not properly document this so far).
```python
import lightly_train

method_args = {
    "teacher": "dinov2_vitg14",
}

if __name__ == "__main__":
    lightly_train.train(
        out="out/my_experiment",            # Output directory
        data="my_data_dir",                 # Directory with images
        model="torchvision/resnet18",       # Model to train
        method_args=method_args,            # Additional arguments for the method
    )
```
You can find the currently available teacher models in https://github.com/lightly-ai/lightly-train/blob/04b48bea58e4e47db07cbdc60cdd86759c654627/src/lightly_train/_modules/teachers/dinov2/build_teacher.py#L24-L41

3. Yes, we are actively working on implementing DINOv2, which will allow SSL pretraining from scratch! Some components are already present inside `src/lightly_train/_methods/dinov2`. This will still take a bit of time, so make sure to get notified when we make a new release!

Hope that answers your questions!

**Comment by DP1701** [comment]
Hi @liopeer ,

Thank you very much for your answer. I have one more question:
Is there a way - like in Ultralytics - to activate an early stop if the network does not improve in the next N epochs?

Following the recommendation in the documentation, I have set the number of epochs between 5,000 and 10,000 for a dataset of 50,000 images. However, training takes a very long time. As I am still new to self-supervised learning, it is difficult for me to judge whether this number of epochs is not too high - especially as there may be no improvement after numerous epochs and I am not sure whether overfitting will occur.

**Comment by liopeer** [comment]
Hi @DP1701!
If you take the `"distillation"` method (which is LightlyTrain's default), you will not experience any overfitting. My recommendation is to train as long as your use case allows or until the loss stops decreasing (but this takes a very long time). We are also actively working on methods that speed up convergence of the models such that training can be sped up in the future.
For contrastive learning (e.g. the `"simclr"` method), effects similar to overfitting might occur where your model doesn't generalize anymore. However it is really difficult in my experience to get into a situation where this happens.

**Comment by DP1701** [comment]
Hi @liopeer ,

Thanks again for the clarification. 

Perhaps you are already aware of this, but I have encountered a problem. For example, if I set num_workers to 120, the training stops after epoch 0, but this does not happen if I set num_workers to 110 (CPU has 128 physical cores).

[log.txt](https://github.com/user-attachments/files/20062956/log.txt)

**Comment by liopeer** [comment]
No problem @DP1701!

About your worker issue: Workers are set *per GPU* and are subprocesses that are spawned off, so creating and communicating between them actually adds some overhead (clearly too much overhead in your case). As I can see from your logs, your machine has 4 GPUs total and you used 2 of them for training. As you'd probably want to use the other 2 GPUs for training as well I would limit the number of workers to 32 (for 4 GPUs and 128 cores total) or 64 (if it has 256 virtual cores). However, the extra overhead created by the processes might actually kill performance.

My recommendation: Start with 8 workers and scale up if you don't reach 100% utlization on your GPUs. In my experience, performance often drops if you go beyond 24 workers.

**Comment by DP1701** [comment]
Hi @liopeer ,

That makes sense – I should have read the documentation more carefully, since that point is mentioned there.

I have another question (sorry for all the questions):

When scaling the images using ```transform_args = {"image_size": (640, 640)}```, is the aspect ratio of the images preserved?

Ultralytics uses padding to reach the target size while keeping the aspect ratio intact.

**Comment by liopeer** [comment]
No problem @DP1701, I am happy to answer your questions!

Under the hood we are using [`RandomResizedCrop` from albumentations](https://albumentations.ai/docs/api-reference/augmentations/crops/transforms/#RandomResizedCrop). This transform first creates a crop of a random aspect ratio (between 0.75 and 1.333) and then resizes the image to "image_size". This means that objects in the image might be slightly squashed/stretched.

Albumentations also has an interactive tool where you can try out the effects: https://explore.albumentations.ai/transform/RandomResizedCrop

Currently it is not possible to configure the `ratio` parameter of `RandomResizedCrop` (which would allow to only allow square crops and therefore not distort anything), but let us know if you'd like to see that exposed to the user.

**Comment by guarin** [comment]
Preservation of the aspect ratio is usually not crucial during pretraining as models are fine-tuned afterwards. In the case of Ultralytics, the model adapts to the fixed aspect ratio during fine-tuning.

**Comment by DP1701** [comment]
Thank you for the clarification! I'm currently wondering whether it makes sense to train using the same input size as Ultralytics—specifically with the yolov11m.pt checkpoint, which was already pre-trained on images with a resolution of 640 px. The same question applies to RF-DETR (Large), which, according to a blog post, was pre-trained with an input size of 728 px.

**Comment by stegmuel** [comment]
Hi @DP1701,
We've had success pre-training on top of Ultralytics weights using the default resolution of 224×224. I’d recommend starting with that and only increasing the resolution if the results don’t meet your expectations, as higher resolutions can be significantly more compute-intensive. In general, we don’t suggest going beyond 518×518, which is the maximum resolution used during the teacher model’s training.

As a side note, our recommended number of epochs is based on training from scratch. Since you're starting from pre-trained weights, you can likely get away with fewer epochs.

**Comment by guarin** [comment]
Will close this for now. Please reopen if you have more questions :)

**Comment by DP1701** [comment]
Hello everyone, I have noticed the new version.
Quick question, the selection of the model for the distillation method has changed a bit.

While before it was ```method_args = {“teacher”: ‘dinov2_vitg14’,}``` now it is ```method_args={ ‘teacher’: “dinov2_vit/vitl14_pretrain”,}```.

Is there a list of possible teacher models that can be used?
This [list](https://github.com/lightly-ai/lightly-train/blob/04b48bea58e4e47db07cbdc60cdd86759c654627/src/lightly_train/_modules/teachers/dinov2/build_teacher.py#L24-L41)  seems to be out of date

**Comment by liopeer** [comment]
Hi @DP1701!

There are indeed changes ongoing to how the teacher model is chosen (which is why we do not yet publicly document how this can be configured). The supported models are now listed here:
https://github.com/lightly-ai/lightly-train/blob/12830783a3a3c9590463d6fd19a4f5e08392b4a5/src/lightly_train/_models/dinov2_vit/dinov2_vit_src/configs/__init__.py#L16-L49

Make sure to use the `_pretrain` suffix and prepend `dinov2_vit/` just like you already realized. And be aware that these things might again change in the future.
